pkg/controller/controllerutil/controllerutil.go

```diff
@@ -168,3 +169,47 @@ func mutate(f MutateFn, key client.ObjectKey, obj runtime.Object) error {
 
 // MutateFn is a function which mutates the existing object into it's desired state.
 type MutateFn func() error
+
+// AddFinalizer accepts a metav1 object and adds the provided finalizer if not present.
+func AddFinalizer(o metav1.Object, finalizer string) {
+	f := o.GetFinalizers()
+	for _, e := range f {
+		if e == finalizer {
+			return
+		}
+	}
+	o.SetFinalizers(append(f, finalizer))
+}
+
+// AddFinalizerIfPossible tries to convert a runtime object to a metav1 object and add the provided finalizer.
+// It returns an error if the provided object cannot provide an accessor.
+func AddFinalizerIfPossible(o runtime.Object, finalizer string) error {
+	m, err := meta.Accessor(o)
+	if err != nil {
+		return err
+	}
+	AddFinalizer(m, finalizer)
+	return nil
+}
+
+// RemoveFinalizer accepts a metav1 object and removes the provided finalizer if present.
+func RemoveFinalizer(o metav1.Object, finalizer string) {
+	f := o.GetFinalizers()
+	for i, e := range f {
+		if e == finalizer {
+			f = append(f[:i], f[i+1:]...)
```

Comment:
Theoretically this is not a stable iteration since you change the slice during the iteration but the index keeps increasing - see this code for example:
```
package main

import "fmt"

func main() {
	a := []int{1, 2, 1, 2}
	for i, v := range a {
		fmt.Printf("a = %+v\n", a)
		fmt.Printf("i = %d\n", i)
		fmt.Printf("v = %d\n", v)
		fmt.Printf("a[i] = %d\n", a[i])
		if a[i] == 1 {
			a = append(a[:i], a[i+1:]...)
		}
	}
}
```
And the output:
```
a = [1 2 1 2]
i = 0
v = 1
a[i] = 1
a = [2 1 2]
i = 1
v = 1
a[i] = 1
a = [2 2]
i = 2
v = 2
panic: runtime error: index out of range
exit status 2
```

How about using the filter example from here: 
https://github.com/golang/go/wiki/SliceTricks#filter-in-place

Reply:
dropped a `return`, but yeah that works too

Reply:
will get it when I squash